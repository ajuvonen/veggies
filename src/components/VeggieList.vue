<script lang="ts" setup>
import {computed} from 'vue';
import {useI18n} from 'vue-i18n';
import {ALL_VEGGIES} from '@/utils/constants';
import {Category, type TranslatedListing} from '@/utils/types';
import {getCategoryForVeggie} from '@/utils/helpers';
import VeggieCompletionChart from '@/components/charts/VeggieCompletionChart.vue';

defineProps<{
  uniqueVeggies: string[];
}>();

const {t, locale} = useI18n();

const allVeggies = computed(() => {
  const collator = new Intl.Collator(locale.value);
  return ALL_VEGGIES.map<TranslatedListing>((veggie) => ({
    veggie,
    category: getCategoryForVeggie(veggie) as Category,
    translation: t(`veggies.${veggie}`),
  })).sort((a, b) => collator.compare(a.translation, b.translation));
});
</script>

<template>
  <div class="veggie-list" data-test-id="veggie-list">
    <VeggieCompletionChart :veggies="uniqueVeggies" />
    <ul v-for="category in Category" :key="category" class="veggie-list__category">
      <ContentElement :title="$t(`categories.${category}`)">
        <ul class="veggie-list__veggies">
          <li
            v-for="{veggie, translation} in allVeggies.filter((item) => item.category === category)"
            :key="veggie"
            class="veggie-list__veggie"
          >
            <IconComponent
              :icon="uniqueVeggies.includes(veggie) ? 'checkboxMarked' : 'checkboxBlank'"
            />
            <span class="truncate capitalize" :title="translation">
              {{ translation }}
            </span>
            <span :data-test-id="`veggie-list-status-${veggie}`" class="sr-only">{{
              $t(uniqueVeggies.includes(veggie) ? 'veggieList.complete' : 'veggieList.missing')
            }}</span>
          </li>
        </ul>
      </ContentElement>
    </ul>
  </div>
</template>
<style scoped>
.veggie-list {
  @apply has-scroll m-0 p-0;
  @apply flex-container flex-col gap-4;
}

.veggie-list__category {
  @apply flex-container gap-4 flex-col;
}

.veggie-list__veggies {
  @apply columns-2 md:columns-3;
}

.veggie-list__veggie {
  @apply flex-container items-center;
}
</style>
