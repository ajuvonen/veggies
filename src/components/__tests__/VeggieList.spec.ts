import {describe, it, expect} from 'vitest';
import {mount} from '@vue/test-utils';
import VeggieList from '@/components/VeggieList.vue';

describe('VeggieList', () => {
  it('renders', () => {
    const wrapper = mount(VeggieList, {
      props: {
        uniqueVeggies: [],
      },
      global: {
        stubs: {
          VeggieCompletionChart: true,
        },
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders with data', () => {
    const uniqueVeggies = ['wheat', 'apple', 'cucumber', 'lychee'];
    const wrapper = mount(VeggieList, {
      props: {
        uniqueVeggies,
      },
      global: {
        stubs: {
          VeggieCompletionChart: true,
        },
      },
    });
    uniqueVeggies.forEach((veggie) => {
      expect(wrapper.findByTestId(`veggie-list-status-${veggie}`).text()).toBe('(complete)');
    });
  });
});
