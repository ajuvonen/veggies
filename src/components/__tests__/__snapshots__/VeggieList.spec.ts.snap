// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VeggieList > renders 1`] = `
"<div data-v-5aea2802="" class="flex min-h-0">
  <div data-v-5aea2802="" class="veggie-list" data-test-id="veggie-list">
    <veggie-completion-chart-stub data-v-5aea2802="" veggies=""></veggie-completion-chart-stub>
    <ul data-v-5aea2802="" class="veggie-list__category">
      <content-element-stub data-v-5aea2802="" title="Fruits And Berries" aria-hidden="false"></content-element-stub>
    </ul>
    <ul data-v-5aea2802="" class="veggie-list__category">
      <content-element-stub data-v-5aea2802="" title="Vegetables" aria-hidden="false"></content-element-stub>
    </ul>
    <ul data-v-5aea2802="" class="veggie-list__category">
      <content-element-stub data-v-5aea2802="" title="Leafy Greens And Herbs" aria-hidden="false"></content-element-stub>
    </ul>
    <ul data-v-5aea2802="" class="veggie-list__category">
      <content-element-stub data-v-5aea2802="" title="Roots And Bulbs" aria-hidden="false"></content-element-stub>
    </ul>
    <ul data-v-5aea2802="" class="veggie-list__category">
      <content-element-stub data-v-5aea2802="" title="Beans And Legumes" aria-hidden="false"></content-element-stub>
    </ul>
    <ul data-v-5aea2802="" class="veggie-list__category">
      <content-element-stub data-v-5aea2802="" title="Grains, Nuts, And Seeds" aria-hidden="false"></content-element-stub>
    </ul>
  </div>
</div>"
`;
